src/compute/src/sink/persist_sink.rs

```diff
@@ -562,6 +567,112 @@ impl<T, R> CorrectionBuffer<T, R> {
     }
 }
 
+/// A stash for storing future updates by time.
+#[derive(Default)]
+struct UpdateStash<D>(BTreeMap<Timestamp, ConsolidatingVec<D>>);
+
+impl<D: Ord> UpdateStash<D> {
+    fn new() -> Self {
+        Self(Default::default())
+    }
+
+    /// Insert a batch of updates into the stash.
+    fn insert(&mut self, mut updates: Vec<(D, Timestamp, Diff)>) {
+        consolidate_updates(&mut updates);
+        updates.sort_unstable_by_key(|(_, time, _)| *time);
+
+        let mut updates = updates.into_iter().peekable();
+        while let Some(&(_, time, _)) = updates.peek() {
+            let data = updates
+                .peeking_take_while(|(_, t, _)| *t == time)
+                .map(|(d, _, r)| (d, r));
+
+            use std::collections::btree_map::Entry;
+            match self.0.entry(time) {
+                Entry::Vacant(entry) => {
+                    entry.insert(data.collect());
+                }
+                Entry::Occupied(mut entry) => {
+                    entry.get_mut().extend(data);
+                }
+            }
+        }
+    }
+
+    /// Remove all updates before the given frontier and return an iterator over them.
+    fn drain_before(
+        &mut self,
+        frontier: &Antichain<Timestamp>,
```

Comment:
Elsewhere we'd use an `AntichainRef`, but no need to change since this is just an internal method.

Reply:
I'm leaving it like it is because it avoids one `.borrow()` at the call site, but something to reconsider if we ever want to use this type more widely.